File: Binary Search/Search Insert Position/searchInsertPosition.go
/* Given a sorted array of distinct integers and a target value, return the index if the target is found. If not, return the index where it would be if it were inserted in order.

You must write an algorithm with O(log n) runtime complexity.

 

Example 1:

Input: nums = [1,3,5,6], target = 5
Output: 2
Example 2:

Input: nums = [1,3,5,6], target = 2
Output: 1
Example 3:

Input: nums = [1,3,5,6], target = 7
Output: 4
 

Constraints:

1 <= nums.length <= 104
-104 <= nums[i] <= 104
nums contains distinct values sorted in ascending order.
-104 <= target <= 104
*/
package main

import (
	"fmt"
)

func searchInsert(nums []int, target int) int {
    var (
        low = 0
        high = len(nums)-1
    )

    for low <= high {
        mid := (low+high)/2
        if nums[mid] == target {
            return mid
        } else if nums[mid] > target {
            high = mid - 1
        } else {
            low = mid+1
        }
    }

    return low
}

func main() {
	fmt.Println(searchInsert([]int{1,3,5,6}, 5))
	fmt.Println(searchInsert([]int{1,3,5,6}, 2))
	fmt.Println(searchInsert([]int{1,3,5,6}, 7))
}

File: Binary Search/Search Insert Position/Search_Insert_Position.cpp
class Solution
{
public:
    int searchInsert(vector<int> &nums, int target)
    {
        // int ans=0;

        // for(int i=0;i<nums.size();i++)
        // {
        //     if(nums[i]==target)
        //     {
        //         ans=i;
        //         break;
        //     }
        //     else if(nums[i]<target)
        //     {
        //         ans=i+1;
        //     }
        //     else
        //     {
        //         break;
        //     }
        // }
        // return ans;
        int n = nums.size();
        int ans;
        int low = 0, high = n - 1;

        while (low <= high)
        {
            int mid = low + (high - low) / 2;

            if (nums[mid] == target)
            {
                ans = mid;
                break;
            }
            else if (nums[mid] < target)
            {
                ans = mid + 1;
                low = mid + 1;
            }
            else
            {
                high = mid - 1;
            }
        }
        return ans;
    }
};